### PR TITLE
Enforce external id uniqueness during DesiredNode construction

### PR DESCRIPTION
**Pull request by fcofdez**
This commit introduces some small refactorings to improve the desired
nodes codebase.

- DesiredNode must contain a valid external id, otherwise it cannot
  be built.
- DesiredNodes now stores desired nodes as a map that uses desired
  nodes external id as the key. This fixes a small bug around
  idempotent updates, as before we were using a list and comparing
  the desired nodes using that list.

### CODE REVIEW

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNode.java [hunk 1]
```diff
@@ -80,13 +85,36 @@ private static Version parseVersion(String version) {
...
+    public DesiredNode(Settings settings, int processors, ByteSizeValue memory, ByteSizeValue storage, Version version) {
+        assert settings != null;
         assert memory != null;
         assert storage != null;
         assert version != null;
         if (processors <= 0) {
             throw new IllegalArgumentException("processors must be greater than 0, but got " + processors);
         }
+
+        if (NODE_EXTERNAL_ID_SETTING.get(settings).isBlank()) {
```
[the remark this answers is not in the document]
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNode.java:105]: We were validating this during the settings validation, but we weren't enforcing it in the DesiredNode constructor, so I think we should be good here. See the rest test `Test external_id or node.name is required`

File: server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java [hunk 2]
```diff
@@ -110,4 +130,49 @@ private static void checkForDuplicatedExternalIDs(List<DesiredNode> nodes) {
...
+        return nodes.values();
+    }
+
+    @Nullable
+    public DesiredNode find(String externalId) {
+        return nodes.get(externalId);
+    }
+
+    private static Map<String, DesiredNode> toMap(final Collection<DesiredNode> desiredNodes) {
+        // use a linked hash map to preserve order
```
[the remark this answers is not in the document]
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DesiredNodes.java:169]: I've moved to an ordered map based on the desired node `external_id` so we get a predictable ordering here.

**Comment by fcofdez** [comment]
Thanks Henning!

**Pull request merged**